Arena: use specialization to avoid copying data

bugadani opened the pull request:
In several cases, a `Vec` or `SmallVec` is passed to `Arena::alloc_from_iter` directly. This PR makes sure those cases don't copy their data unnecessarily, by specializing the `alloc_from_iter` implementation.

rust-highfive commented:
r? @varkor

(rust_highfive has picked a reviewer for you, use r? to override)

bugadani commented:
cc @est31

varkor commented:
Apologies for taking so long; this slipped under my radar. I'm not familiar with this code, so I'm assigning to @Mark-Simulacrum based on `git blame`.

r? @Mark-Simulacrum

Mark-Simulacrum commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
+#[inline]
+unsafe fn write_from_iter<'a, T, I: Iterator<Item = T>>(
+    mut iter: I,
+    len: usize,
```
Can we call this "expected_len" or something like that? Otherwise it seems a bit confusing to see a len that's only a hint, really.

Mark-Simulacrum commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
 }
 
+#[inline]
+unsafe fn write_from_iter<'a, T, I: Iterator<Item = T>>(
```
Can we add a comment on the unsafe here, i.e., what are the preconditions necessary?

bugadani replied:
Actually, extracting this function is not necessary because of the upcoming changes prompted by https://github.com/rust-lang/rust/pull/78569#discussion_r526506964

Mark-Simulacrum commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
+            vec.as_ptr().copy_to_nonoverlapping(start_ptr, len);
+            vec.set_len(0);
+            slice::from_raw_parts_mut(start_ptr, len)
+        }
```
Is there a reason this is duplicating the specialization for SmallVec? I would expect us to just forward to that here.

Mark-Simulacrum commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
+        unsafe {
+            let len = self.len();
+            let start_ptr = arena.alloc_raw_slice(len);
+            write_from_iter(self, len, start_ptr)
```
Hm, len() here is exact, right, so calling write_from_iter which is more general feels a bit off? We should probably be doing a direct memcpy here?

(It looks like `std::array::IntoIter` does not yet expose a as_slice function or similar, but that seems like something we can just add in this PR, unstably? cc @scottmcm who I seem to recall was doing some of the array iter work?)

scottmcm replied:
The `slice::Iter` has `.as_slice()` and is `AsRef<[T]>`, so having those on `array::IntoIter` seems like non-controversial pattern following, so that sounds good to me.  (Please take with appropriate salt considering I'm not on libs, but I would personally be willing to r+ it as unstable.)

est31 replied:
The slice function takes a copy though while the purpose of IntoIter is to do a move. Ideally we wouldn't do a copy but directly write into the arena, but I guess that won't work because the array construction might call the arena itself, and it might also panic.

est31 replied:
Made a PR: #79194

bugadani commented:
@Mark-Simulacrum I've done what I could. I've addressed your comments and reverted some changes based on those comments. For example, changes to `write_from_iter` are no longer part of this PR, as they aren't necessary any more.

Mark-Simulacrum commented:
@bors try @rust-timer queue

This should just be a win (or roughly neutral), but let's check. r=me otherwise.

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 43b83b334130ef066d8518ca2a125779399ccd73 with merge aa21b48c4d75183274998a446bae14e98053637a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"43b83b334130ef066d8518ca2a125779399ccd73","merge_sha":"aa21b48c4d75183274998a446bae14e98053637a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1432441503)
Build commit: aa21b48c4d75183274998a446bae14e98053637a (`aa21b48c4d75183274998a446bae14e98053637a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1432441503"},"merge_sha":"aa21b48c4d75183274998a446bae14e98053637a"} -->

rust-timer commented:
Queued aa21b48c4d75183274998a446bae14e98053637a with parent c9c57fadc47c8ad986808fc0a47479f6d2043453, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c9c57fadc47c8ad986808fc0a47479f6d2043453&end=aa21b48c4d75183274998a446bae14e98053637a).

rust-timer commented:
Finished benchmarking try commit (aa21b48c4d75183274998a446bae14e98053637a): [comparison url](https://perf.rust-lang.org/compare.html?start=c9c57fadc47c8ad986808fc0a47479f6d2043453&end=aa21b48c4d75183274998a446bae14e98053637a).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

bugadani commented:
Interesting. I wonder if I see the results correctly - an increase in instruction counts, but in the same time, looking at the detailed results, a decrease in times? For example, this report belongs to a 0.5% instruction count regression, but shows a -0.3% total time diff: https://perf.rust-lang.org/detailed-query.html?commit=aa21b48c4d75183274998a446bae14e98053637a&base_commit=c9c57fadc47c8ad986808fc0a47479f6d2043453&benchmark=match-stress-enum-check&run_name=incr-unchanged

Hmm.

bugadani commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +118,72 @@ impl<T> Default for TypedArena<T> {
+impl<T, const N: usize> IterExt<T> for std::array::IntoIter<T, N> {
+    #[inline]
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        if N == 0 {
```
I think this should be `self.len() == 0` or similar.

Mark-Simulacrum replied:
The code below should work with a zero length, right? Unless alloc_from_slice has a nonzero precondition on the length which isn't documented...

bugadani replied:
It should. My thought was that the iterator is not guaranteed to be on the first element, so `N == self.len()` isn't guaranteed. I shall doule check later, thoguh.

Mark-Simulacrum replied:
Aha, no, the code below will not work -- alloc_raw_slice asserts a nonzero length. So we should compare against the length here.

bugadani replied:
Fixed, and I've aligned the other places as well.

Mark-Simulacrum commented:
I would say that this is perhaps a slight regression in instruction counts, but overall a win on cycles -- https://perf.rust-lang.org/compare.html?start=c9c57fadc47c8ad986808fc0a47479f6d2043453&end=aa21b48c4d75183274998a446bae14e98053637a&stat=cycles%3Au on non-incremental benchmarks. I am inclined to land it based on these performance results; it does not seem like an obvious win yet, but it does seem like something we should be doing to try and avoid potential problems for the cases optimized here.

Mark-Simulacrum commented:
Ok, r=me with commits squashed

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit e93a4637c0858725a0d7fdebeecf49bd97add724 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum e93a4637c0858725a0d7fdebeecf49bd97add724 -->
<!-- homu: {"type":"Approved","sha":"e93a4637c0858725a0d7fdebeecf49bd97add724","approver":"Mark-Simulacrum"} -->

est31 commented:
Are the cycle numbers always this noisy? There is a 5% on incr-unchanged on the encoding-check benchmark. Apparently most of the slowdown comes from `early_lint_checks` being 50% slower. https://perf.rust-lang.org/detailed-query.html?commit=aa21b48c4d75183274998a446bae14e98053637a&base_commit=c9c57fadc47c8ad986808fc0a47479f6d2043453&benchmark=encoding-check&run_name=incr-unchanged

Should this worry us?

Mark-Simulacrum commented:
Note that the 50% there is a wall clock measurement of 6ms, which is almost certainly noise. Cycle measurements frequently have noise in the 5% range on clean benchmarks; for incremental unchanged benchmarks they can be even noisier since there's less overall cycles on those.

I am not particularly worried.

est31 commented:
@Mark-Simulacrum thanks for the info!

bors commented:
:hourglass: Testing commit e93a4637c0858725a0d7fdebeecf49bd97add724 with merge 432d116a5c2565774bae4c42fcacab8b685608b5...
<!-- homu: {"type":"BuildStarted","head_sha":"e93a4637c0858725a0d7fdebeecf49bd97add724","merge_sha":"432d116a5c2565774bae4c42fcacab8b685608b5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1434418272)
Approved by: Mark-Simulacrum
Pushing 432d116a5c2565774bae4c42fcacab8b685608b5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1434418272"},"merge_sha":"432d116a5c2565774bae4c42fcacab8b685608b5"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#78569!

Tested on commit rust-lang/rust@432d116a5c2565774bae4c42fcacab8b685608b5.
Direct link to PR: <https://github.com/rust-lang/rust/pull/78569>

💔 rls on linux: test-pass → test-fail (cc @Xanewok).

pickfire commented on compiler/rustc_arena/src/lib.rs:
```diff
@@ -114,6 +118,72 @@ impl<T> Default for TypedArena<T> {
+            mem::forget(self);
+            slice::from_raw_parts_mut(start_ptr, len)
+        }
+    }
```
Looks like there are quite a bit of code duplication here, shouldn't it be refactored?

bugadani replied:
Well, it's not intentionally duplicated, but I couldn't find a way to make it significantly nicer. Unfortunately, we can't use `alloc_from_slice` because it requires the items to be `Copy`.
